src/adapter/src/optimize/dataflows.rs

```diff
@@ -206,7 +206,7 @@ impl<'a> DataflowBuilder<'a> {
                 let entry = self.catalog.get_entry(id);
                 match entry.item() {
                     CatalogItem::Table(table) => {
-                        dataflow.import_source(*id, table.desc.typ().clone(), false);
+                        dataflow.import_source(*id, table.desc.typ().clone(), table.insert_only);
```

Comment:
The correctness here is contingent on the rest of the system not allowing negative diffs in the results of queries. This is afaiu informally true but various opinions exist about where it should need to be enforced (e.g. in storage, in persist, in compute, etc).  I'm not sure this is necessarily more likely to happen than elsewhere, but elsewhere it is a bug when we announce a thing as append-only and then don't meet that contract, vs here where it may just be laundering those bugs. I *think* there's the ability to actually verify that the writes into the table do not contain negative differences, though I can't easily point out where that is. I think it's probably worth putting something in the vicinity of that work, to understand if the contract is being violated or not.

Reply:
> I _think_ there's the ability to actually verify that the writes into the table do not contain negative differences, though I can't easily point out where that is. I think it's probably worth putting something in the vicinity of that work, to understand if the contract is being violated or not.

That makes sense to me too. I think `Coordinator::send_diffs`, where the transaction writeops get constructed, is one reasonable place to perform this check; alternatively, `Coordinator::sequence_end_transaction`, where the transaction writeops actually get submitted. These places both have the advantage that we could return an error to the end user rather than panicking, but the disadvantage that they are not perfectly protective. By contrast, putting a check in `Coordinator::group_commit_initiate` would be perfectly protective, but I think it is much harder to gracefully fail a user write from in there.

Reply:
> here where it may just be laundering those bugs

In addition to (rather than instead of) what Nikhil said, would it make sense to add a check for non-negative diffs in/after the source operator in rendering? It seems like this would be pretty cheap, in that we wouldn't have to consolidate anything.

Reply:
Independently from the above discussion, `monotonic_view_inner`'s `CatalogItem::Table` case should also be modified, so that we take into account the monotonicity of such tables that are used through indexes or views.